pkg/controller/direct/privilegedaccessmanager/entitlement_controller.go

```diff
@@ -0,0 +1,442 @@
+			"'obj.Spec.PrivilegedAccess.GcpIAMAccess.OrganizationRef': %w", err)
+	}
+	switch *obj.Spec.PrivilegedAccess.GcpIAMAccess.ResourceType {
+	case "cloudresourcemanager.googleapis.com/Project":
+		if !strings.HasPrefix(container, "projects/") {
+			return nil, fmt.Errorf("only 'spec.privilegedAccess.gcpIAMAccess.projectRef' "+
+				"should be configured because the corresponding resourceType is "+
+				"'cloudresourcemanager.googleapis.com/Project', but got resource %s", container)
+		}
+		obj.Spec.PrivilegedAccess.GcpIAMAccess.ProjectRef.External = container
+	case "cloudresourcemanager.googleapis.com/Folder":
+		if !strings.HasPrefix(container, "folders/") {
+			return nil, fmt.Errorf("only 'spec.privilegedAccess.gcpIAMAccess.folderRef' "+
+				"should be configured because the corresponding resourceType is "+
+				"'cloudresourcemanager.googleapis.com/Folder', but got resource %s", container)
+		}
+		obj.Spec.PrivilegedAccess.GcpIAMAccess.FolderRef.External = container
+	case "cloudresourcemanager.googleapis.com/Organization":
+		if !strings.HasPrefix(container, "organizations/") {
+			return nil, fmt.Errorf("only 'spec.privilegedAccess.gcpIAMAccess.organizationRef' "+
+				"should be configured because the corresponding resourceType is "+
+				"'cloudresourcemanager.googleapis.com/Organization', but got resource %s", container)
+		}
+		obj.Spec.PrivilegedAccess.GcpIAMAccess.OrganizationRef.External = container
+	default:
+		return nil, fmt.Errorf("unrecoganizable resourceType: %v; must be one of "+
+			"'cloudresourcemanager.googleapis.com/Project', "+
+			"'cloudresourcemanager.googleapis.com/Folder', "+
+			"'cloudresourcemanager.googleapis.com/Organization'",
+			*obj.Spec.PrivilegedAccess.GcpIAMAccess.ResourceType)
+	}
+
+	if obj.Spec.RequesterJustificationConfig.NotMandatory == nil && obj.Spec.RequesterJustificationConfig.Unstructured == nil {
+		return nil, fmt.Errorf("one and only one of 'spec.requesterJustificationConfig.notMandatory' " +
+			"and 'spec.requesterJustificationConfig.unstructured' should be configured: neither is configured")
+	}
+	if obj.Spec.RequesterJustificationConfig.NotMandatory != nil && obj.Spec.RequesterJustificationConfig.Unstructured != nil {
+		return nil, fmt.Errorf("one and only one of 'spec.requesterJustificationConfig.notMandatory' " +
+			"and 'spec.requesterJustificationConfig.unstructured' should be configured: both configured")
+	}
+
+	// Get privilegedaccessmanager GCP client
+	gcpClient, err := m.client(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return &Adapter{
+		id:        id,
+		gcpClient: gcpClient,
+		desired:   obj,
+	}, nil
+}
+
+func oneOfContainer(ctx context.Context, reader client.Reader, obj *krm.PrivilegedAccessManagerEntitlement, projectRef *refs.ProjectRef, folderRef *refs.FolderRef, organizationRef *refs.OrganizationRef) (string, error) {
+	container := ""
+	if projectRef == nil && folderRef == nil && organizationRef == nil {
```

Comment:
Nit: it might be simpler just to count the number of non-nils.  I think your code is correct though, but it's easy to imagine it becoming incorrect once ten people have changed this function

Reply:
Will refactor it in https://github.com/GoogleCloudPlatform/k8s-config-connector/pull/2659.